src/distributors/base-installer.ts

```diff
@@ -0,0 +1,144 @@
+import * as tc from '@actions/tool-cache';
+import * as core from '@actions/core';
+import semver from 'semver';
+import path from 'path';
+import * as httpm from '@actions/http-client';
+import {  getVersionFromToolcachePath, parseLocalVersions } from '../util';
+
+export interface JavaInstallerOptions {
+    version: string;
+    arch: string;
+    javaPackage: string;
+}
+
+export interface IJavaInfo {
+    javaVersion: string;
+    javaPath: string;
+}
+
+export interface IJavaRelease {
+    resolvedVersion: string;
+    link: string;
+}
+
+export abstract class JavaBase {
+    protected http: httpm.HttpClient;
+    protected version: string;
+    protected arch: string;
+    protected javaPackage: string;
+    constructor(protected distributor: string, initOptions: JavaInstallerOptions) {
+        this.http = new httpm.HttpClient('setup-java', undefined, {
+            allowRetries: true,
+            maxRetries: 3
+          });
+          this.version = this.normalizeVersion(initOptions.version);
+          this.arch = initOptions.arch;
+          this.javaPackage  = initOptions.javaPackage;
+    }
+
+    protected abstract downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo>;
+    protected abstract resolveVersion(range: semver.Range): Promise<IJavaRelease>;
+    protected abstract get javaRootName(): string;
+
+    public async getJava(): Promise<IJavaInfo> {
+        const range = new semver.Range(this.version);
+        let foundJava = this.findInToolcache(range);
+        if (!foundJava) {
+            // try to find Java in default system locations outside tool-cache
+            foundJava = this.findInKnownLocations(range)
```

Comment:
Let's remove findInKnownLocations and related logic and move Java to toolcache on images

Reply:
Done.